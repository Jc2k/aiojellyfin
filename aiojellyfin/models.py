"""Jellyfin API models."""

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Generic, TypeVar

from mashumaro.config import BaseConfig
from mashumaro.mixins.json import DataClassJSONMixin
from mashumaro.types import Discriminator

from .const import ImageType, ItemType, MediaProtocol, MediaSourceType


class JellyfinConfig(BaseConfig):
    """Mashumaro configuration."""

    forbid_extra_keys = False


@dataclass(kw_only=True)
class MediaStream(DataClassJSONMixin):
    """Information about a Jellyfin stream."""

    class Config(BaseConfig):
        """Mashumaro configuration."""

        discriminator = Discriminator(
            field="Type",
            include_subtypes=True,
        )

        aliases = {
            "codec": "Codec",
            "time_base": "TimeBase",
            "audio_spacial_format": "AudioSpacialFormat",
            "is_interlaced": "IsInterlaced",
            "is_avc": "IsAVC",
            "is_default": "IsDefault",
            "is_forced": "IsForced",
            "is_hearing_impaired": "IsHearingImpaired",
            "profile": "Profile",
            "media_stream_type": "Type",
            "index": "Index",
            "is_external": "IsExternal",
            "is_text_subtitle_stream": "IsTextSubtitleStream",
            "supports_external_stream": "SupportsExternalStream",
            "level": "Level",
        }

    codec: str
    time_base: str | None
    audio_spacial_format: str | None = None
    is_interlaced: bool | None
    is_avc: bool | None
    is_default: bool | None
    is_forced: bool | None
    is_hearing_impaired: bool | None
    profile: str | None = None
    media_stream_type: str | None
    index: int | None
    is_external: bool | None
    is_text_subtitle_stream: bool | None
    supports_external_stream: bool | None
    level: bool | None


@dataclass(kw_only=True)
class AudioMediaStream(MediaStream):
    """An audio media stream."""

    Type = "Audio"

    CodecTag: str | None = None
    Language: str | None = None
    DisplayTitle: str
    ChannelLayout: str
    BitRate: int
    Channels: int
    SampleRate: int


@dataclass(kw_only=True)
class EmbeddedImageMediaStream(MediaStream):
    """An embedded image stream."""

    Type = "EmbeddedImage"


@dataclass(kw_only=True)
class MediaSource(DataClassJSONMixin):
    """Information about a Jellyfin media source."""

    Config = JellyfinConfig

    Protocol: MediaProtocol
    Id: str | None
    Path: str
    Type: MediaSourceType
    Container: str | None
    Size: int | None
    Name: str | None
    SupportsDirectPlay: bool | None
    SupportsDirectStream: bool | None
    SupportsTranscoding: bool | None
    MediaStreams: Sequence[MediaStream] | None


@dataclass(kw_only=True)
class ArtistItem:
    """Information about a relationship between media and an artist."""

    Config = JellyfinConfig

    Id: str
    Name: str


@dataclass(kw_only=True)
class UserData:
    """Metadata that is specific to the logged in user, like favorites."""

    Config = JellyfinConfig

    IsFavorite: bool


@dataclass(kw_only=True)
class MediaLibrary:
    """JSON data describing a single media library."""

    Config = JellyfinConfig

    Id: str
    Name: str
    CollectionType: str


@dataclass(kw_only=True)
class MediaLibraries:
    """JSON data describing a collection of media libraries."""

    Config = JellyfinConfig

    Items: list[MediaLibrary]
    TotalRecordCount: int
    StartIndex: int


@dataclass(kw_only=True)
class MediaItem(DataClassJSONMixin):
    """JSON data describing a single media item."""

    class Config(BaseConfig):
        """Config for mashumaro."""

        discriminator = Discriminator(
            field="Type",
            include_subtypes=True,
        )

        aliases = {
            "id": "Id",
            "item_type": "Type",
            "name": "Name",
            "media_type": "MediaType",
            "sort_name": "SortName",
            "provider_ids": "ProviderIds",
            "run_time_ticks": "RunTimeTicks",
            "image_tags": "ImageTags",
            "backdrop_image_tags": "BackdropImageTags",
            "user_data": "UserData",
            "overview": "Overview",
            "index_number": "IndexNumber",
        }

    id: str
    item_type: ItemType
    name: str
    media_type: str
    sort_name: str | None = None
    provider_ids: dict[str, str] | None = None
    run_time_ticks: int
    image_tags: dict[ImageType, str]
    backdrop_image_tags: list[str]
    user_data: UserData | None = None
    overview: str | None = None
    index_number: int | None = None


MediaItemT = TypeVar("MediaItemT", bound=MediaItem)


@dataclass(kw_only=True)
class MediaItems(Generic[MediaItemT], DataClassJSONMixin):
    """JSON data describing a collection of media items."""

    Config = JellyfinConfig

    Items: list[MediaItemT]
    TotalRecordCount: int
    StartIndex: int


@dataclass(kw_only=True)
class Artist(MediaItem):
    """JSON data describing a single artist."""

    Type = ItemType.MusicArtist


@dataclass(kw_only=True)
class Album(MediaItem):
    """JSON data describing a single album."""

    Type = ItemType.MusicAlbum

    CanDownload: bool | None = None
    ProductionYear: int | None = None
    ArtistItems: list[ArtistItem]
    AlbumArtists: list[ArtistItem]


@dataclass(kw_only=True)
class Track(MediaItem):
    """JSON data describing a single track."""

    Type = ItemType.Audio

    AlbumId: str | None = None
    Album: str | None = None
    AlbumArtist: str | None = None
    CanDownload: bool | None = None
    ProductionYear: int | None = None
    ParentIndexNumber: int
    MediaSources: list[MediaSource] | None = None
    MediaStreams: Sequence[MediaStream] | None = None
    NormalizationGain: int | None = None
    ArtistItems: list[ArtistItem]
    AlbumArtists: list[ArtistItem]


@dataclass(kw_only=True)
class Playlist(MediaItem):
    """JSON data describing a single playlist."""

    Type = ItemType.Playlist
