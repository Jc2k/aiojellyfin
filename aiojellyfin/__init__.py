import uuid
from aiohttp import ClientSession
from typing import LiteralString, TypedDict, Required, cast
import urllib

class MediaLibrary(TypedDict, total=False):
    Id: Required[str]
    Name: Required[str]
    CollectionType: str

class MediaLibraries(TypedDict):
    Items: list[MediaLibrary]
    TotalRecordCount: int
    StartIndex: 0

class Artist(TypedDict, total=False):
    Id: Required[str]
    Name: Required[str]

class Artists(TypedDict):
    Items: list[MediaLibrary]
    TotalRecordCount: int
    StartIndex: 0


"""
search_media_items
get_item
artwork
"""

class Connection:

    def __init__(self, url: str, user_id: str, device_id: str, access_token: str):
        self.base_url = url
        self._session = ClientSession(
            base_url=url,
        )
        self._user_id = user_id
        self._access_token = access_token
        self._device_id = device_id

    async def _get_json(self, url, params):
        resp = await self._session.get(
            url,
            params=params,
            headers={
                "Content-Type": "application/json",
                "User-Agent": "Music Assistant/0.0.0",
                "Authorization": _get_authenication_header("Music Assistant", "Music Assistant", str(uuid.uuid4()), "0.0.0", self._access_token),
            },
            raise_for_status=True,
        )
        return await resp.json()
    
    async def get_media_folders(self, fields=None) -> MediaLibraries:
        params = {}
        if fields:
            params["fields"] = fields
        resp = await self._get_json("/Items", params=params)
        return cast(MediaLibraries, resp)
    
    async def artists(self, library_id: str) -> Artists:
        resp = await self._get_json(
            "/Artists",
            params={
                "ParentId": library_id,
                "ArtistType": "Artist,AlbumArtist",
            },
        )
        return cast(Artists, resp)
    
    async def user_items(self, handler: LiteralString="", params=None):
        # FIXME: This will be removed ASAP with something with more typing
        return await self._get_json(
            f"/Items{handler}",
            params=params,
        ) 

    def audio_url(self, item_id: str, container=None, audio_codec=None, max_streaming_bitrate=140000000) -> str:
        params = {
            "UserId": self._user_id,
            "DeviceId": self._device_id,
            "MaxStreamingBitrate": max_streaming_bitrate,
        }

        if container:
            params["Container"] = container

        if audio_codec:
            params["AudioCodec"] = audio_codec

        if "api_key" not in params:
            params["api_key"] = self._access_token

        encoded = urllib.parse.urlencode(params)

        payload = f"{self.base_url}Audio/{item_id}/universal?{encoded}"
        
        return payload
    

def _get_authenication_header(name: str, device: str, device_id: str, version: str, token: str=None) -> str:
    params = {
        "Client": name,
        "Device": device,
        "DeviceId": device_id,
        "Version": version
    }
    if token:
        params["Token"] = token
    param_line = ", ".join(f'{k}="{v}"' for k, v in params.items())
    return f"MediaBrowser {param_line}"


async def authenticate_by_name(url: str, username: str, password: str="") -> Connection:
    session = ClientSession(
        base_url=url,
    )
    async with session:
        res = await session.post(
            "/Users/AuthenticateByName",
            json={"Username": username, "Pw": password},
            headers={
                "Content-Type": "application/json",
                "User-Agent": "Music Assistant/0.0.0",
                "Authorization": _get_authenication_header("Music Assistant", "Music Assistant", str(uuid.uuid4()), "0.0.0"),
            },
            raise_for_status=True,
        )
        user_session = await res.json()

    user = user_session["User"]
    session_info = user_session["SessionInfo"]

    return Connection(url, user["Id"], session_info["DeviceId"], user_session["AccessToken"])