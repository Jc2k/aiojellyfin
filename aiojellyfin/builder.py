"""Builder pattern for building Jellyfin API queries."""

import asyncio
import copy
from collections.abc import AsyncGenerator
from typing import Generic, Self

from mashumaro.codecs.basic import BasicDecoder

from .const import ItemFields, ItemType
from .models import Album, Artist, MediaItems, MediaItemT, Playlist, Track
from .session import Session


class ItemQueryBuilder(Generic[MediaItemT]):
    """A query that can be performed against /Items or similar Jellyfin endpoint."""

    endpoint = "/Items"

    _params: dict[str, str]
    _decoder: BasicDecoder[MediaItems[MediaItemT]]

    def __init__(
        self,
        session: Session,
        decoder: BasicDecoder[MediaItems[MediaItemT]],
        params: dict[str, str],
    ):
        """Initialise the class."""
        self._decoder = decoder
        self._session = session
        self._params = params

    @classmethod
    def create(cls, session: Session) -> Self:
        """Create a new query builder instance."""
        decoder = BasicDecoder(MediaItems[MediaItemT])
        return cls(session, decoder, {})

    def _clone(self) -> Self:
        return self.__class__(
            self._session,
            self._decoder,
            copy.deepcopy(self._params),
        )

    def parent(self, parent_id: str) -> Self:
        """Specify this to localize the search to a specific item or folder.

        Omit to use the root.
        """
        result = self._clone()
        result._params["parentId"] = parent_id
        return result

    def search_term(self, search_term: str) -> Self:
        """Search for this term."""
        result = self._clone()
        result._params["searchTerm"] = search_term
        return result

    def include_item_types(self, *args: ItemType) -> Self:
        """Only include these item types."""
        result = self._clone()
        result._params["includeItemTypes"] = ",".join(arg.value for arg in args)
        return result

    def exclude_item_types(self, *args: ItemType) -> Self:
        """Exclude these item types."""
        result = self._clone()
        result._params["excludeItemTypes"] = ",".join(args)
        return result

    def recursive(self, recursive: bool) -> Self:
        """Search parent folder and all child folders."""
        result = self._clone()
        result._params["recursive"] = "true" if recursive else "false"
        return result

    def enable_userdata(self) -> Self:
        """Include per user metadata - does the logged in user like the content."""
        result = self._clone()
        result._params["enableUserData"] = "true"
        return result

    def fields(self, *args: ItemFields) -> Self:
        """Specify additional fields of information to return in the output."""
        result = self._clone()
        result._params["fields"] = ",".join(field.value for field in args)
        return result

    def start_index(self, start_index: int) -> Self:
        """Record index to start at.

        All items with a lower index will be dropped from the results.
        """
        result = self._clone()
        result._params["startIndex"] = str(start_index)
        return result

    def limit(self, limit: int) -> Self:
        """Maximum number of records to return."""
        result = self._clone()
        result._params["limit"] = str(limit)
        return result

    def to_params(self) -> dict[str, str]:
        """Build a dictionary of the query parameters for this search."""
        return copy.deepcopy(self._params)

    async def request(self) -> MediaItems[MediaItemT]:
        """Request a list of records matching this query."""
        response = await self._session.get_json(self.endpoint, params=self.to_params())
        return self._decoder.decode(response)

    async def stream(self, page_size: int = 100) -> AsyncGenerator[MediaItemT, None]:
        """Stream all records matching this query with automatic greedy pagination."""
        request = self.limit(page_size)
        response = await request.request()
        offset = 0

        while offset < response["TotalRecordCount"]:
            offset += page_size
            next_response = asyncio.create_task(request.start_index(offset).request())

            for obj in response["Items"]:
                yield obj

            response = await next_response

        for obj in response["Items"]:
            yield obj


class ArtistQueryBuilder(ItemQueryBuilder[Artist]):
    """Builder for searching artist records."""

    endpoint = "/Artists"

    @classmethod
    def setup(cls, session: Session) -> Self:
        """Initialise the builder with default search restrictions for artists."""
        return super().create(session).recursive(True)


class AlbumQueryBuilder(ItemQueryBuilder[Album]):
    """Builder for searching album records."""

    @classmethod
    def setup(cls, session: Session) -> Self:
        """Initialise the builder with default search restrictions for albums."""
        return super().create(session).include_item_types(ItemType.MusicAlbum).recursive(True)


class TrackQueryBuilder(ItemQueryBuilder[Track]):
    """Builder for searching track records."""

    @classmethod
    def setup(cls, session: Session) -> Self:
        """Initialise the builder with default search restrictions for tracks."""
        return super().create(session).include_item_types(ItemType.Audio).recursive(True)


class PlaylistQueryBuilder(ItemQueryBuilder[Playlist]):
    """Builder for searching playlist records."""

    @classmethod
    def setup(cls, session: Session) -> Self:
        """Initialise the builder with default search restrictions for playlists."""
        return super().create(session).include_item_types(ItemType.Playlist).recursive(True)
