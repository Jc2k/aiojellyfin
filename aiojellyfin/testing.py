"""Helpers for testing users of aiojellyfin."""

import copy
import json
from collections import defaultdict
from collections.abc import Awaitable, Callable, Mapping
from typing import Any

from aiojellyfin import Connection, NotFound, session
from aiojellyfin.models import (
    Album,
    Artist,
    MediaItems,
    MediaLibraries,
    Playlist,
    Track,
)
from aiojellyfin.session import SessionConfiguration

MUSIC_FOLDER = "8aeb9430-b1d5-420e-9847-3217ac2120c3"


class FixtureBuilder:
    """A set of test data for testing against."""

    def __init__(self) -> None:
        """Init the class."""
        self.objects: dict[str, Any] = {}
        self.objects_parents: dict[str, set[str]] = defaultdict(set)

    def add_json_bytes(self, data: str | bytes) -> None:
        """Add an artist to this fixture."""
        fixture = json.loads(data)
        fixture_id = fixture["Id"]
        self.objects[fixture_id] = fixture
        self.objects_parents[fixture_id].add(MUSIC_FOLDER)

        if parents := fixture.get("AlbumArtists"):
            for artist in parents:
                self.objects_parents[fixture_id].add(artist["Id"])
        if parents := fixture.get("ArtistItems"):
            for artist in parents:
                self.objects_parents[fixture_id].add(artist["Id"])
        if album_id := fixture.get("AlbumId"):
            self.objects_parents[fixture_id].add(album_id)

    def to_authenticate_by_name(
        self,
    ) -> Callable[[SessionConfiguration, str, str], Awaitable[Connection]]:
        """Prepare a monkeypatch."""

        async def authenticate_by_name(
            session_config: SessionConfiguration, _username: str, _password: str
        ) -> Connection:
            return TestConnection(session_config, "TestUserId", "TestAccessToken", self)

        return authenticate_by_name


class Session(session.Session):
    """A fake session for testing only."""

    def __init__(
        self,
        session_config: SessionConfiguration,
        user_id: str,
        access_token: str,
        fixture: FixtureBuilder,
    ):
        """Init the fake session."""
        self.session_config = session_config
        self.user_id = user_id
        self.access_token = access_token
        self.fixture = fixture

    async def get_json(self, url: str, params: Mapping[str, str]) -> dict[str, Any]:
        """Generate a fake response from fixtures."""
        if url in ("/Items", "/Artists"):
            results = []

            for record_id, record in self.fixture.objects.items():
                if url == "/Artists" and record.get("Type") != "MusicArtist":
                    continue

                if parent_id := params.get("parentId"):
                    if parent_id not in self.fixture.objects_parents[record_id]:
                        continue

                if search_term := params.get("searchTerm"):
                    if search_term.lower() not in record["Name"].lower():
                        continue

                if item_types_raw := params.get("includeItemTypes"):
                    item_types = item_types_raw.split(",")
                    if record["Type"] not in item_types:
                        continue

                record_copy = copy.deepcopy(record)

                if not params.get("enableUserData"):
                    record_copy.pop("UserData", None)

                # if fields:

                results.append(record)

            total_record_count = len(results)

            if start_index := params.get("startIndex"):
                results = results[int(start_index) :]

            if limit := params.get("limit"):
                results = results[: int(limit)]

            return {
                "Items": results,
                "StartIndex": start_index or 0,
                "TotalRecordCount": total_record_count,
            }

        raise RuntimeError("Unfaked http request detected")


class TestConnection(Connection):
    """A fake implementation of Connection."""

    def __init__(
        self,
        session_config: SessionConfiguration,
        user_id: str,
        access_token: str,
        fixture: FixtureBuilder,
    ) -> None:
        """Init the class."""
        super().__init__(session_config, user_id, access_token)
        self._fixture = fixture
        self._session = Session(session_config, user_id, access_token, fixture)
        self.artists._session = self._session
        self.albums._session = self._session
        self.tracks._session = self._session
        self.playlists._session = self._session

    async def get_media_folders(self, fields: str | None = None) -> MediaLibraries:
        """Fetch a list of media libraries."""
        return {
            "Items": [
                {
                    "Id": MUSIC_FOLDER,
                    "Name": "Music",
                    "CollectionType": "music",
                }
            ],
            "TotalRecordCount": 1,
            "StartIndex": 0,
        }

    async def get_artist(self, artist_id: str) -> Artist:
        """Fetch all data for a single artist."""
        try:
            return self._fixture.objects[artist_id]
        except KeyError:
            raise NotFound(artist_id)

    async def get_album(self, album_id: str) -> Album:
        """Fetch all data for a single album."""
        try:
            return self._fixture.objects[album_id]
        except KeyError:
            raise NotFound(album_id)

    async def get_track(self, track_id: str) -> Track:
        """Fetch all data for a single track."""
        try:
            return self._fixture.objects[track_id]
        except KeyError:
            raise NotFound(track_id)

    async def get_playlist(self, playlist_id: str) -> Playlist:
        """Fetch all data for a single playlist."""
        try:
            return self._fixture.objects[playlist_id]
        except KeyError:
            raise NotFound(playlist_id)

    async def get_suggested_tracks(self) -> MediaItems[Track]:
        """Return suggested tracks."""
        return {
            "Items": [next(iter(self._fixture.tracks.values()))],
            "StartIndex": 0,
            "TotalRecordCount": 1,
        }

    async def get_similar_tracks(
        self,
        track_id: str,
        limit: int | None = None,
        fields: list[str] | None = None,
    ) -> MediaItems[Track]:
        """Return similar tracks."""
        return {
            "Items": [next(iter(self._fixture.tracks.values()))],
            "StartIndex": 0,
            "TotalRecordCount": 1,
        }
