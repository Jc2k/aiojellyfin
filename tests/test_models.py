"""Test we can parse Jellyfin models into Music Assistant models."""

import pathlib

import pytest
from syrupy.assertion import SnapshotAssertion

from aiojellyfin.models import Album, Artist, Track

FIXTURES_DIR = pathlib.Path(__file__).parent / "fixtures"
ARTIST_FIXTURES = list(FIXTURES_DIR.glob("artists/*.json"))
ALBUM_FIXTURES = list(FIXTURES_DIR.glob("albums/*.json"))
TRACK_FIXTURES = list(FIXTURES_DIR.glob("tracks/*.json"))


@pytest.mark.parametrize("example", ARTIST_FIXTURES, ids=lambda val: str(val.stem))
def test_parse_artists(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse artists."""
    with open(example) as fp:
        parsed = Artist.from_json(fp.read())
    assert snapshot == parsed


@pytest.mark.parametrize("example", ALBUM_FIXTURES, ids=lambda val: str(val.stem))
def test_parse_albums(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse albums."""
    with open(example) as fp:
        parsed = Album.from_json(fp.read())
    assert snapshot == parsed


@pytest.mark.parametrize("example", TRACK_FIXTURES, ids=lambda val: str(val.stem))
def test_parse_tracks(example: pathlib.Path, snapshot: SnapshotAssertion) -> None:
    """Test we can parse tracks."""
    with open(example) as fp:
        parsed = Track.from_json(fp.read())
    assert snapshot == parsed
