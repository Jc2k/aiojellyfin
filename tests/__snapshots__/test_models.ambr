# serializer version: 1
# name: test_parse_albums[infest]
  Album(id='70b7288088b42d318f75dbcc41fd0091', item_type=<ItemType.MusicAlbum: 'MusicAlbum'>, name='Infest', media_type='Unknown', sort_name='infest', provider_ids={'MusicBrainzAlbum': 'bf25b030-0cbb-495a-8d79-6c7fee20a089', 'MusicBrainzReleaseGroup': '0193355a-cdfb-3936-afd2-44d651eb006d', 'MusicBrainzAlbumArtist': 'c5eb9407-caeb-4303-b383-6929aa94021c'}, run_time_ticks=27614273019, image_tags={<ImageType.Primary: 'Primary'>: 'bcbe1ac159b0522743c9a0fe5401f948'}, backdrop_image_tags=[], user_data=UserData(IsFavorite=False), overview=None, index_number=None, CanDownload=False, ProductionYear=2000, ArtistItems=[ArtistItem(Id='e439648e08ade14e27d5de48fa97c88e', Name='Papa Roach')], AlbumArtists=[ArtistItem(Id='e439648e08ade14e27d5de48fa97c88e', Name='Papa Roach')])
# ---
# name: test_parse_albums[this_is_christmas]
  Album(id='32ed6a0091733dcff57eae67010f3d4b', item_type=<ItemType.MusicAlbum: 'MusicAlbum'>, name='This Is Christmas', media_type='Unknown', sort_name='this is christmas', provider_ids={'MusicBrainzAlbum': 'b13a174d-527d-44a1-b8f8-a4c78b03b7d9', 'MusicBrainzReleaseGroup': 'f002d6b7-17af-4f9e-8d30-5486548ffe6f', 'MusicBrainzAlbumArtist': '60bbceb2-0ddc-403b-970b-b4e9c3b2de5c/827b9ff1-56f8-4614-9261-a08de5fc1be0'}, run_time_ticks=18722017229, image_tags={<ImageType.Primary: 'Primary'>: 'b685ba2b9247aca1ea66dda557bb8f54'}, backdrop_image_tags=[], user_data=UserData(IsFavorite=False), overview=None, index_number=None, CanDownload=None, ProductionYear=2011, ArtistItems=[ArtistItem(Id='a0c459294295710546c81c20a8d9abfc', Name='Emmy the Great'), ArtistItem(Id='1952db245ddef4e41dcd016475379190', Name='Tim Wheeler')], AlbumArtists=[ArtistItem(Id='555b36f7d310d1b7405557a8775c6878', Name='Emmy the Great & Tim Wheeler')])
# ---
# name: test_parse_albums[yesterday_when_i_was_mad]
  Album(id='7c8d0bd55291c7fc0451d17ebef30017', item_type=<ItemType.MusicAlbum: 'MusicAlbum'>, name='Yesterday, When I Was Mad [Disc 2]', media_type='Unknown', sort_name='yesterday when i was mad [disc 0000000002]', provider_ids={}, run_time_ticks=0, image_tags={}, backdrop_image_tags=[], user_data=UserData(IsFavorite=False), overview=None, index_number=None, CanDownload=None, ProductionYear=None, ArtistItems=[], AlbumArtists=[])
# ---
# name: test_parse_artists[ash]
  Artist(id='dd954bbf54398e247d803186d3585b79', item_type=<ItemType.MusicArtist: 'MusicArtist'>, name='Ash', media_type='Unknown', sort_name='ash', provider_ids={'MusicBrainzArtist': '99164692-c02d-407c-81c9-25d338dd21f4'}, run_time_ticks=509234691363, image_tags={<ImageType.Primary: 'Primary'>: '8a543e58fda6d2f374263a4dcd0d2fbd', <ImageType.Logo: 'Logo'>: '662f82868ad2964190daea171f3fcf08'}, backdrop_image_tags=['8a4c3c67629b28673de7af433a1efd68'], user_data=UserData(IsFavorite=False), overview=None, index_number=None)
# ---
# name: test_parse_tracks[thrown_away]
  Track(id='b5319fb11cde39fca2023184fcfa9862', item_type=<ItemType.Audio: 'Audio'>, name='11 Thrown Away', media_type='Audio', sort_name='0000 - 0000 - 11 Thrown Away', provider_ids={}, run_time_ticks=5777763270, image_tags={}, backdrop_image_tags=[], user_data=UserData(IsFavorite=False), overview=None, index_number=0, AlbumId='70b7288088b42d318f75dbcc41fd0091', Album=None, AlbumArtist=None, CanDownload=True, ProductionYear=None, ParentIndexNumber=0, MediaSources=[MediaSource(Protocol=<MediaProtocol.File: 'File'>, Id='b5319fb11cde39fca2023184fcfa9862', Path='/media/music/Papa Roach/Infest/11 Thrown Away.m4a', Type=<MediaSourceType.Default: 'Default'>, Container='mp3', Size=11283443, Name='11 Thrown Away', SupportsDirectPlay=True, SupportsDirectStream=True, SupportsTranscoding=True, MediaStreams=[AudioMediaStream(codec='mp3', time_base='1/14112000', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=False, is_forced=False, is_hearing_impaired=False, profile=None, media_stream_type='Audio', index=0, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=0, CodecTag=None, Language=None, DisplayTitle='MP3 - Stereo', ChannelLayout='stereo', BitRate=156231, Channels=2, SampleRate=44100)])], MediaStreams=[AudioMediaStream(codec='mp3', time_base='1/14112000', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=False, is_forced=False, is_hearing_impaired=False, profile=None, media_stream_type='Audio', index=0, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=0, CodecTag=None, Language=None, DisplayTitle='MP3 - Stereo', ChannelLayout='stereo', BitRate=156231, Channels=2, SampleRate=44100)], NormalizationGain=None, ArtistItems=[], AlbumArtists=[])
# ---
# name: test_parse_tracks[where_the_bands_are]
  Track(id='54918f75ee8f6c8b8dc5efd680644f29', item_type=<ItemType.Audio: 'Audio'>, name='Where the Bands Are (2018 Version)', media_type='Audio', sort_name='0001 - 0001 - Where the Bands Are (2018 Version)', provider_ids={}, run_time_ticks=2464333790, image_tags={<ImageType.Primary: 'Primary'>: 'dbd792d6c27313d01ed7c2dce85f785b'}, backdrop_image_tags=[], user_data=UserData(IsFavorite=False), overview=None, index_number=1, AlbumId=None, Album='Where the Bands Are (2018 Version) - Single', AlbumArtist='Dead Like Harry', CanDownload=True, ProductionYear=2018, ParentIndexNumber=1, MediaSources=[MediaSource(Protocol=<MediaProtocol.File: 'File'>, Id='54918f75ee8f6c8b8dc5efd680644f29', Path='/media/music/Dead Like Harry/01 Where the Bands Are (2018 Version).m4a', Type=<MediaSourceType.Default: 'Default'>, Container='m4a', Size=9167268, Name='01 Where the Bands Are (2018 Version)', SupportsDirectPlay=True, SupportsDirectStream=True, SupportsTranscoding=True, MediaStreams=[AudioMediaStream(codec='aac', time_base='1/44100', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=True, is_forced=False, is_hearing_impaired=False, profile='LC', media_stream_type='Audio', index=0, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=0, CodecTag='mp4a', Language='eng', DisplayTitle='English - AAC - Stereo - Default', ChannelLayout='stereo', BitRate=278038, Channels=2, SampleRate=44100), EmbeddedImageMediaStream(codec='mjpeg', time_base='1/90000', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=False, is_forced=False, is_hearing_impaired=False, profile='Baseline', media_stream_type='EmbeddedImage', index=1, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=-99)])], MediaStreams=[AudioMediaStream(codec='aac', time_base='1/44100', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=True, is_forced=False, is_hearing_impaired=False, profile='LC', media_stream_type='Audio', index=0, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=0, CodecTag='mp4a', Language='eng', DisplayTitle='English - AAC - Stereo - Default', ChannelLayout='stereo', BitRate=278038, Channels=2, SampleRate=44100), EmbeddedImageMediaStream(codec='mjpeg', time_base='1/90000', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=False, is_forced=False, is_hearing_impaired=False, profile='Baseline', media_stream_type='EmbeddedImage', index=1, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=-99)], NormalizationGain=-10, ArtistItems=[ArtistItem(Id='94875b0dd58cbf5245a135982133651a', Name='Dead Like Harry')], AlbumArtists=[ArtistItem(Id='94875b0dd58cbf5245a135982133651a', Name='Dead Like Harry')])
# ---
# name: test_parse_tracks[zombie_christmas]
  Track(id='fb12a77f49616a9fc56a6fab2b94174c', item_type=<ItemType.Audio: 'Audio'>, name='Zombie Christmas', media_type='Audio', sort_name='0001 - 0008 - Zombie Christmas', provider_ids={'MusicBrainzArtist': '60bbceb2-0ddc-403b-970b-b4e9c3b2de5c/827b9ff1-56f8-4614-9261-a08de5fc1be0', 'MusicBrainzAlbumArtist': '60bbceb2-0ddc-403b-970b-b4e9c3b2de5c/827b9ff1-56f8-4614-9261-a08de5fc1be0', 'MusicBrainzAlbum': 'b13a174d-527d-44a1-b8f8-a4c78b03b7d9', 'MusicBrainzReleaseGroup': 'f002d6b7-17af-4f9e-8d30-5486548ffe6f', 'MusicBrainzTrack': '17d1019d-d4f4-326c-b4bb-d8aec2607bd7'}, run_time_ticks=2249317010, image_tags={<ImageType.Primary: 'Primary'>: 'c8e39ff125c3ba39a5791570dffa4b83'}, backdrop_image_tags=[], user_data=UserData(IsFavorite=False), overview=None, index_number=8, AlbumId='32ed6a0091733dcff57eae67010f3d4b', Album='This Is Christmas', AlbumArtist='Emmy the Great & Tim Wheeler', CanDownload=True, ProductionYear=2011, ParentIndexNumber=1, MediaSources=[MediaSource(Protocol=<MediaProtocol.File: 'File'>, Id='fb12a77f49616a9fc56a6fab2b94174c', Path='/media/music/Emmy the Great & Tim Wheeler/This Is Christmas/8. Zombie Christmas.m4a', Type=<MediaSourceType.Default: 'Default'>, Container='m4a', Size=8225981, Name='8. Zombie Christmas', SupportsDirectPlay=True, SupportsDirectStream=True, SupportsTranscoding=True, MediaStreams=[AudioMediaStream(codec='aac', time_base='1/44100', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=True, is_forced=False, is_hearing_impaired=False, profile='LC', media_stream_type='Audio', index=0, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=0, CodecTag='mp4a', Language='eng', DisplayTitle='English - AAC - Stereo - Default', ChannelLayout='stereo', BitRate=267933, Channels=2, SampleRate=44100), EmbeddedImageMediaStream(codec='mjpeg', time_base='1/90000', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=False, is_forced=False, is_hearing_impaired=False, profile='Baseline', media_stream_type='EmbeddedImage', index=1, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=-99)])], MediaStreams=[AudioMediaStream(codec='aac', time_base='1/44100', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=True, is_forced=False, is_hearing_impaired=False, profile='LC', media_stream_type='Audio', index=0, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=0, CodecTag='mp4a', Language='eng', DisplayTitle='English - AAC - Stereo - Default', ChannelLayout='stereo', BitRate=267933, Channels=2, SampleRate=44100), EmbeddedImageMediaStream(codec='mjpeg', time_base='1/90000', audio_spacial_format=None, is_interlaced=False, is_avc=False, is_default=False, is_forced=False, is_hearing_impaired=False, profile='Baseline', media_stream_type='EmbeddedImage', index=1, is_external=False, is_text_subtitle_stream=False, supports_external_stream=False, level=-99)], NormalizationGain=-11, ArtistItems=[ArtistItem(Id='a0c459294295710546c81c20a8d9abfc', Name='Emmy the Great'), ArtistItem(Id='1952db245ddef4e41dcd016475379190', Name='Tim Wheeler')], AlbumArtists=[ArtistItem(Id='555b36f7d310d1b7405557a8775c6878', Name='Emmy the Great & Tim Wheeler')])
# ---
